src/libstd/sys_common/gnu/libbacktrace.rs

```diff
@@ -121,12 +123,27 @@ pub fn print(w: &mut Write, idx: isize, addr: *mut libc::c_void,
     // case. There is no evidence at the moment to suggest that a more carefully
     // constructed file can't cause arbitrary code execution. As a result of all
     // of this, we don't hint libbacktrace with the path to the current process.
-    unsafe fn init_state() -> *mut backtrace_state {
+    //
+    // This can also return a File which should remain open until
+    // the first use of the state.
+    unsafe fn init_state() -> (*mut backtrace_state, Option<fs::File>) {
         static mut STATE: *mut backtrace_state = ptr::null_mut();
-        if !STATE.is_null() { return STATE }
-        STATE = backtrace_create_state(ptr::null(), 0, error_cb,
+        if !STATE.is_null() { return (STATE, None) }
+
+        let (filename, file) = match ::sys::backtrace::get_executable_filename() {
+            Ok((filename, file)) => {
+                // filename is purposely leaked here since libbacktrace requires
+                // it to stay allocated permanently
+                let filename_ptr = filename.as_ptr();
+                mem::forget(filename);
+                (filename_ptr, Some(file))
+            },
+            Err(_) => (ptr::null(), None),
+        };
+
+        STATE = backtrace_create_state(filename, 0, error_cb,
                                        ptr::null_mut());
-        STATE
+        (STATE, file)
```

Comment:
We only need to keep the file open/locked during the first call to a libbacktrace function that operates on the state unless that function fails and as such it will try the initialization again on the next call.

What is currently done is to optionally return the file on the first creation of the state. And if the first call to `libbacktrace_syminfo` fails we leak it so it stays open, otherwise it will be dropped on exit from `print`.

The options I can think of:
1. Refactor the code so it's clearer, I'm open to any suggestions here.
2. Leak it always. `Just mem::forget` immediately.
3. Return it and keep it in a `static mut` (Annoying since it's a drop type so we need to play around with `Box` and raw pointers), and drop it explicitly once `backtrace_syminfo` succeeds the first time.
4. The same as 3 but place it in a `static mut` immediately in `init_state`.